### DISCUSSION TITLE
Snappy v2

### DISCUSSION THREAD

**Discussion by AntoineLelaisant** [KnpLabs/snappy#467] [Announcements]
Hi guys!

We are currently looking forward to rewrite snappy in order to make it more flexible and able to use other backends such as headless chrome, SaaS solutions, etc.

## Motivations

- [wkhtmltopdf](https://github.com/wkhtmltopdf/wkhtmltopdf) has been abandoned on January 2023. (also see #466)
- There is plenty of tools allowing to generate pdf from html nowadays. Our vision for Snappy is to have more flexibility to easily switch from one tool to another.
- We also want to provide a more flexible way of producing pdfs. Forcing the user to generate a file on the filesystem is not always the best solution. We want to provide a way to stream the output and let the user decide where to store it afterward.

## Big picture about the new approach

Snappy is currently built on top of two main classes: `Pdf` and `Image`. We plan to replace them by a root `Snappy` class that would provide a unified method to generate a pdf from an uri or an html string. The method signature would be something like:

```php
public function generate(string $input, array $options = []): StreamInterface
```

Behind the scene, the `Snappy` class would use an adapter pattern to generate the pdf. The adapter will be responsible for actually generating the pdf (or the image).

The final usage would be something like:

```php
$adapter = new WkhtmlToPdfAdapter();

$snappy = new Snappy($adapter);

$stream = $snappy->generate('https://knplabs.com');
```

## Adapters

We plan to provide the following adapters:
    - `Wkhtmltopdf` (already implemented)
    - `Wkhtmltoimage` (already implemented)
    - `Chrome headless` (new)

Note for `wkhtmltopdf`: Even if the project is abandoned we will continue to maintain an adapter for it. This should ease the process of migrating from Snappy v1 to Snappy v2.

Nevertheless, we will provide and document a way to create your own adapter. Indeed, we want to make it easy to use other tools such as SaaS solutions.

## Foreseen Breaking Changes

- PHP support: We will drop the support for php `<=v8.0`. (see: <https://www.php.net/supported-versions.php>)
- Namespace: `Knp\Snappy` will be replaced by `KnpLabs\Snappy` in order to be consistent with the other KnpLabs FOSS projects.
- Snappy will no longer produce a file on the filesystem. Instead, it will return a stream of the output.
- Snappy will no longer accept multiple inputs to generate a single pdf. Indeed `wkhtmltopdf` used to provide this feature but this does not seem to exist on most other tools.

## Feedbacks

Feel free to comment this discussion if you have any feedbacks or ideas about the new approach. All of these changes are still a plan and we are open to discuss about it. We are looking forward to hear from you!

**Comment by AntoineLelaisant** [comment]
Here is the draft PR for the v2 <https://github.com/KnpLabs/snappy/pull/468> (_still in progress_)

**Comment by PedroTroller** [comment]
FYI https://github.com/KnpLabs/snappy/issues/462

**Comment by PedroTroller** [comment]
@AntoineLelaisant it is also the occasion to drop support of deprecated versions of `symfony/process`

**Comment by PedroTroller** [comment]
For me, a good solution is to have for example
```php
Snappy\HtmlToPdf::generate(string $html, array $options = []): StreamInterface
Snappy\UriToPdf::generate(UriInterface $uri, array $options = []): StreamInterface
Snappy\FileToPdf::generate(SplFileInfo $file, array $options = []): StreamInterface
```
All of these are plugged to adapters and there adapters must implements one or more of interfaces
```php
Snappy\Backend\HtmlToPdf
Snappy\Backend\UriToPdf
Snappy\Backend\FileToPdf
```
And for example, if you use a tranformer named `chrome`  with class `Snappy\Adapter\CromeHeadless` which only implements `Snappy\Adapter\UriToPdfTransformer`  then the SnappyBundle will create only one service in the DIC with declaration `Snappy\UriToPdf $chrome` . But if you use a transformer named `wkhtmltopdf` with class `Snappy\Adapter\WkhtmlToPdf`  which implements all three interfaces, then the bundle will mount 3 declatations,
- `Snappy\Frontend\HtmlToPdf $wkhtmltopdf`
- `Snappy\Frontend\UriToPdf $wkhtmltopdf`
- `Snappy\Frontend\FileToPdf $wkhtmltopdf`

**Reply by baiiko** [reply]
Why don't you go through a Provider? which will trigger a service according to the argument passed as, for example, the Serializer symfony does with its support method?

**Reply by PedroTroller** [reply]
And the argument supported is the uri, html or file ?

**Reply by baiiko** [reply]
In the method support you check the type of $value and you accept or no to execute. Look Serializer process on symfony. 

Like a stretegy pattern wrapped with a provider who call support method :)

**Comment by rlataguerra** [comment]
You said you want to make it more flexible, even with things like SaaS solution ?
What do you exactly have in mind ? I am asking because with the deprecation of wkhtmltopdf and how hard puppeteer can be to install or manage, I ended up creating a small SaaS (doppio.sh)

I would be interested to contribute to the SaaS integration (somekind of backend plugin system ?)

**Reply by alexpozzi** [reply]
I will write you the moment we will have finalized the backends interface so that you'll be able to work on your own backend integration.
At the moment, that interface is not finalized yet.

**Comment by Chris53897** [comment]
Ideas sounds good. Is there any news on this?

**Reply by alexpozzi** [reply]
Unfortunately the advancement is slow but steady.
Lately we are lacking of time to dedicate to open source projects.

**Reply by vasilvestre** [reply]
Is there any way to have tickets to help you ?

**Comment by DavidPetrasek** [comment]
Are you sure chrome headless is gonna work also on shared hosting? 
I tried running chrome-php on shared hosting (here: https://www.webglobe.com/webhosting - START variant) and it's not possible to make it start.
It works only on my local server.

here is the output: 

```
[2023-10-12T00:03:18.616208+02:00] vlastni.DEBUG: process: starting process: '/home/html/.../public_html/www/bin/chrome-linux64/chrome' '--remote-debugging-port=0' '--remote-allow-origins=*' '--disable-background-networking' '--disable-background-timer-throttling' '--disable-client-side-phishing-detection' '--disable-hang-monitor' '--disable-popup-blocking' '--disable-prompt-on-repost' '--disable-sync' '--disable-translate' '--disable-features=ChromeWhatsNewUI' '--metrics-recording-only' '--no-first-run' '--safebrowsing-disable-auto-update' '--enable-automation' '--password-store=basic' '--use-mock-keychain' '--headless' '--disable-gpu' '--font-render-hinting=none' '--hide-scrollbars' '--mute-audio' '--user-data-dir=/home/html/.../tmp/chromium-php-mVyMbB' [] {"file":"/home/html/.../public_html/www/vendor/chrome-php/chrome/src/Browser/BrowserProcess.php","line":131,"class":"HeadlessChromium\\Browser\\BrowserProcess","callType":"->","function":"start"}
[2023-10-12T00:03:18.620558+02:00] vlastni.DEBUG: process: waiting for 30 seconds for startup [] {"file":"/home/html/.../public_html/www/vendor/chrome-php/chrome/src/Browser/BrowserProcess.php","line":407,"class":"HeadlessChromium\\Browser\\BrowserProcess","callType":"->","function":"waitForStartup"}
[2023-10-12T00:03:18.631649+02:00] vlastni.DEBUG: process: ✗ chrome process stopped [] {"file":"/home/html/.../public_html/www/vendor/chrome-php/chrome/src/Browser/BrowserProcess.php","line":414,"class":"HeadlessChromium\\Browser\\BrowserProcess","callType":"->","function":"HeadlessChromium\\Browser\\{closure}"}
[2023-10-12T00:03:18.719671+02:00] vlastni.DEBUG: process: killing chrome [] {"file":"/home/html/.../public_html/www/vendor/chrome-php/chrome/src/Browser/BrowserProcess.php","line":188,"class":"HeadlessChromium\\Browser\\BrowserProcess","callType":"->","function":"kill"}
```

**Reply by alexpozzi** [reply]
Hello,
We might consider to add an adapter to let you connect to an existing Chrome instance/pool of instances but we need to investigate more on that topic.
Every suggestion is welcome.

**Reply by DavidPetrasek** [reply]
I've also asked about my issue on chrome-php: https://github.com/chrome-php/chrome/issues/555

**Reply by DavidPetrasek** [reply]
The solution is to use VPS, because some providers won't let you execute custom binaries.

**Comment by vasilvestre** [comment]
How's the v2 going ?

**Reply by alexpozzi** [reply]
https://github.com/KnpLabs/snappy/discussions/467#discussioncomment-7261759

**Comment by KevinArtus** [comment]
Hi guys

@alexpozzi @AntoineLelaisant @PedroTroller 
Have you been able to progress on v2 of snappy for the new generation of pdfs without Wkhtmltopdf ?

We have a problem with the version of the pdfs which is generated (1.4) by Wkhtmltopdf. This library does not allow you to generate files in 1.7.
Snapy v2 should fix the problem I think with new ways of generating pdfs. Do you have a release date planned?

**Reply by alexpozzi** [reply]
Hello @KevinArtus!

I personally don't have much time to dedicate to this project lately. 

Can you create an issue about `This library does not allow you to generate files in 1.7.` as I'm not aware of any issue on that side.

**Comment by guideloince** [comment]
What's up here ? 

Any news for improvements ? 
Can you switch to new binary easily like @baiiko say [here](https://github.com/KnpLabs/snappy/issues/456).
